File: source/dynamical_systems/test/tests/test_circular.cpp
#include <gtest/gtest.h>

#include "dynamical_systems/DynamicalSystemFactory.hpp"
#include "dynamical_systems/exceptions/EmptyBaseFrameException.hpp"
#include "dynamical_systems/exceptions/EmptyAttractorException.hpp"

#include "state_representation/exceptions/EmptyStateException.hpp"
#include "state_representation/exceptions/IncompatibleReferenceFramesException.hpp"
#include "state_representation/geometry/Ellipsoid.hpp"
#include "state_representation/parameters/Parameter.hpp"

using namespace dynamical_systems;
using namespace state_representation;
using namespace std::literals::chrono_literals;

class CircularDSTest : public testing::Test {
protected:
  void SetUp() override {
    ds = CartesianDynamicalSystemFactory::create_dynamical_system(DYNAMICAL_SYSTEM_TYPE::CIRCULAR);
    current_pose = CartesianPose("A", 10 * Eigen::Vector3d::Random());
    center = CartesianPose::Identity("B");
    limit_cycle.set_center_pose(center);
    limit_cycle.set_axis_lengths({radius, radius});
  }

  std::shared_ptr<IDynamicalSystem<CartesianState>> ds;
  Ellipsoid limit_cycle = Ellipsoid("limit_cycle");
  CartesianPose current_pose;
  CartesianPose center;
  double radius = 10;
  unsigned int nb_steps = 2000;
  std::chrono::milliseconds dt = 10ms;
  double linear_tol = 1e-3;
};

TEST_F(CircularDSTest, TestPositionOnRadius) {
  ds->set_parameter_value("limit_cycle", limit_cycle);

  for (unsigned int i = 0; i < nb_steps; ++i) {
    CartesianTwist twist = ds->evaluate(current_pose);
    twist.clamp(10, 10, 0.001, 0.001);
    current_pose += dt * twist;
  }

  EXPECT_NEAR(current_pose.dist(center, CartesianStateVariable::POSITION), radius, linear_tol);
}

TEST_F(CircularDSTest, EmptyConstructor) {
  // base frame and attractor should be empty
  EXPECT_TRUE(ds->get_parameter_value<Ellipsoid>("limit_cycle").get_center_pose().is_empty());
  EXPECT_TRUE(ds->get_base_frame().is_empty());

  ds->set_parameter_value("limit_cycle", limit_cycle);
  EXPECT_FALSE(ds->get_parameter_value<Ellipsoid>("limit_cycle").get_center_pose().is_empty());
  EXPECT_FALSE(ds->get_base_frame().is_empty());
  // when attractor was set without a base frame, expect base frame to be identity with name / reference_frame of attractor
  EXPECT_EQ(ds->get_base_frame().get_name(), center.get_reference_frame());
  EXPECT_EQ(ds->get_base_frame().get_reference_frame(), center.get_reference_frame());
  EXPECT_EQ(ds->get_base_frame().get_transformation_matrix(), Eigen::Matrix4d::Identity());
}

TEST_F(CircularDSTest, EvaluateCompatibility) {
  CartesianState state1 = CartesianState::Identity("world", "A");
  CartesianState state2("D", "C");
  CartesianState state3("C", "A");
  CartesianState state4("C", "world");
  // if no base frame is set, an exception is thrown
  EXPECT_THROW(ds->evaluate(state1), dynamical_systems::exceptions::EmptyBaseFrameException);
  ds->set_base_frame(state1);
  // if cartesian state is incompatible, an exception is thrown
  EXPECT_THROW(ds->evaluate(state2), state_representation::exceptions::IncompatibleReferenceFramesException);
  // if cartesian state needs to be transformed to other frame first and is empty, an exception is thrown
  EXPECT_THROW(ds->evaluate(state3), state_representation::exceptions::EmptyStateException);
  // if no attractor is set, an exception is thrown
  EXPECT_THROW(ds->evaluate(state4), dynamical_systems::exceptions::EmptyAttractorException);

  ds->set_parameter_value("limit_cycle", limit_cycle);
  EXPECT_TRUE(ds->is_compatible(state1));
  EXPECT_FALSE(ds->is_compatible(state2));
  EXPECT_TRUE(ds->is_compatible(state3));
  EXPECT_TRUE(ds->is_compatible(state4));
}

TEST_F(CircularDSTest, TestPositionOnRadiusRandomCenter) {
  limit_cycle.set_center_position(Eigen::Vector3d::Random());
  limit_cycle.set_center_orientation(Eigen::Quaterniond::UnitRandom());
  ds->set_parameter_value("limit_cycle", limit_cycle);

  for (unsigned int i = 0; i < nb_steps; ++i) {
    state_representation::CartesianTwist twist = ds->evaluate(current_pose);
    twist.clamp(10, 10, 0.001, 0.001);
    current_pose += dt * twist;
  }

  EXPECT_NEAR(current_pose.dist(limit_cycle.get_center_pose(), CartesianStateVariable::POSITION), radius, linear_tol);
}

TEST_F(CircularDSTest, SetCenterAndBase) {
  Ellipsoid cycle("B", "A");
  auto BinA = CartesianState::Identity("B", "A");
  auto CinA = CartesianState::Identity("C", "A");
  auto CinB = CartesianState::Identity("C", "B");

  ds->set_parameter_value("limit_cycle", cycle);
  EXPECT_STREQ(ds->get_parameter_value<Ellipsoid>("limit_cycle").get_center_pose().get_name().c_str(),
               BinA.get_name().c_str());
  EXPECT_STREQ(ds->get_parameter_value<Ellipsoid>("limit_cycle").get_center_pose().get_reference_frame().c_str(),
               BinA.get_reference_frame().c_str());

  // evaluating a state is only valid if it matches the base reference frame
  EXPECT_NO_THROW(ds->evaluate(CinA));
  EXPECT_THROW(ds->evaluate(CinB), state_representation::exceptions::IncompatibleReferenceFramesException);

  // setting the center is only valid if it matches the base reference frame
  cycle.set_center_state(CinA);
  EXPECT_NO_THROW(ds->set_parameter_value("limit_cycle", cycle));
  Ellipsoid cycle2("C", "B");
  EXPECT_THROW(ds->set_parameter_value("limit_cycle", cycle2),
               state_representation::exceptions::IncompatibleReferenceFramesException);

  // setting the base frame should also update the reference frame of the center
  ASSERT_NO_THROW(ds->set_base_frame(CinB));
  EXPECT_STREQ(ds->get_base_frame().get_name().c_str(), CinB.get_name().c_str());
  EXPECT_STREQ(ds->get_base_frame().get_reference_frame().c_str(), CinB.get_reference_frame().c_str());
  EXPECT_STREQ(ds->get_parameter_value<Ellipsoid>("limit_cycle").get_center_pose().get_reference_frame().c_str(),
               ds->get_base_frame().get_name().c_str());

  // now the base frame is C, setting a center should only work if the reference frame of the center is also C
  EXPECT_THROW(ds->set_parameter_value("limit_cycle", cycle),
               state_representation::exceptions::IncompatibleReferenceFramesException);
  EXPECT_NO_THROW(ds->set_parameter_value("limit_cycle", cycle2));
  EXPECT_NO_THROW(ds->set_parameter_value("limit_cycle", Ellipsoid("B", "C")));
}

File: python/test/dynamical_systems/test_circular.py
import numpy as np
import state_representation as sr
import unittest
from datetime import timedelta
from dynamical_systems import create_cartesian_ds, DYNAMICAL_SYSTEM_TYPE


class TestCircular(unittest.TestCase):

    @classmethod
    def setUp(cls):
        cls.center = sr.CartesianPose.Identity("B")
        cls.limit_cycle = sr.Ellipsoid("limit_cycle")
        cls.radius = 10.0
        cls.nb_steps = 2000
        cls.dt = timedelta(milliseconds=10)
        cls.tol = 1e-3
        cls.limit_cycle.set_center_pose(cls.center)
        cls.limit_cycle.set_axis_lengths([cls.radius, cls.radius])

    def assert_np_array_equal(self, a: np.array, b: np.array, places=3):
        try:
            np.testing.assert_almost_equal(a, b, decimal=places)
        except AssertionError as e:
            self.fail(f'{e}')

    def test_empty_constructor(self):
        ds = create_cartesian_ds(DYNAMICAL_SYSTEM_TYPE.CIRCULAR)
        self.assertTrue(ds.get_parameter_value("limit_cycle").get_center_state().is_empty())
        self.assertTrue(ds.get_base_frame().is_empty())

        ds.set_parameter(sr.Parameter("limit_cycle", self.limit_cycle, sr.ParameterType.STATE, sr.StateType.GEOMETRY_ELLIPSOID))
        self.assertFalse(ds.get_parameter_value("limit_cycle").get_center_state().is_empty())
        self.assertFalse(ds.get_base_frame().is_empty())

        self.assertEqual(ds.get_base_frame().get_name(), self.center.get_reference_frame())
        self.assertEqual(ds.get_base_frame().get_reference_frame(), self.center.get_reference_frame())
        self.assert_np_array_equal(ds.get_base_frame().get_transformation_matrix(), np.eye(4))

    def test_is_compatible(self):
        ds = create_cartesian_ds(DYNAMICAL_SYSTEM_TYPE.CIRCULAR)
        state1 = sr.CartesianState.Identity("world", "A")
        state2 = sr.CartesianState("D", "C")
        state3 = sr.CartesianState("C", "A")
        state4 = sr.CartesianState("C", "world")

        with self.assertRaises(RuntimeError):
            ds.evaluate(state1)

        ds.set_base_frame(state1)
        with self.assertRaises(RuntimeError):
            ds.evaluate(state2)
        with self.assertRaises(RuntimeError):
            ds.evaluate(state3)
        with self.assertRaises(RuntimeError):
            ds.evaluate(state4)

        ds.set_parameter(sr.Parameter("limit_cycle", self.limit_cycle, sr.ParameterType.STATE, sr.StateType.GEOMETRY_ELLIPSOID))
        self.assertTrue(ds.is_compatible(state1))
        self.assertFalse(ds.is_compatible(state2))
        self.assertTrue(ds.is_compatible(state3))
        self.assertTrue(ds.is_compatible(state4))

    def test_points_on_radius_random_center(self):
        ds = create_cartesian_ds(DYNAMICAL_SYSTEM_TYPE.CIRCULAR)
        self.limit_cycle.set_center_position(np.random.rand(3, 1))
        ds.set_parameter(sr.Parameter("limit_cycle", self.limit_cycle, sr.ParameterType.STATE, sr.StateType.GEOMETRY_ELLIPSOID))

        current_pose = sr.CartesianPose("A", 10 * np.random.rand(3, 1))
        for i in range(self.nb_steps):
            twist = sr.CartesianTwist(ds.evaluate(current_pose))
            twist.clamp(10, 10, 0.001, 0.001)
            current_pose += self.dt * twist

        self.assertTrue(current_pose.dist(self.limit_cycle.get_center_pose(),
                                          sr.CartesianStateVariable.POSITION) - self.radius < self.tol)


if __name__ == '__main__':
    unittest.main()
